File: android-native/app/src/main/java/com/mariusschober/goalflow/nativeapp/data/CounterLedger.kt
package com.mariusschober.goalflow.nativeapp.data

import org.json.JSONArray
import org.json.JSONObject
import java.math.BigDecimal
import java.math.BigInteger
import java.time.Instant
import java.time.LocalDate
import java.time.format.DateTimeFormatterBuilder

/** JSON-domain validation shared by the private causal-action ledger. */
internal object ActionJson {
    private val uuid = Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")
    val instantFormatter = DateTimeFormatterBuilder().appendInstant(3).toFormatter()
    const val MAX_SAFE_INTEGER = 9007199254740991L
    fun identity(value: Any?): Boolean = value is String && uuid.matches(value)
    fun integer(value: Any?): Long? {
        if (value !is Number) return null
        return runCatching {
            val decimal = BigDecimal(value.toString())
            val exact = decimal.longValueExact()
            exact.takeIf { it in -MAX_SAFE_INTEGER..MAX_SAFE_INTEGER }
        }.getOrNull()
    }
    fun day(value: Any?): Boolean = value is String && Regex("^\\d{4}-\\d{2}-\\d{2}$").matches(value)
        && runCatching { LocalDate.parse(value).toString() == value }.getOrDefault(false)
    fun instant(value: Any?): Boolean = value is String
        && Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{3}Z$").matches(value)
        && runCatching { instantFormatter.format(Instant.parse(value)) == value }.getOrDefault(false)
    fun canonical(value: Any?): String = when (value) {
        null, JSONObject.NULL -> "null"
        is JSONObject -> value.keys().asSequence().toList().sorted().joinToString(",", "{", "}") { key ->
            JSONObject.quote(key) + ":" + canonical(value.get(key))
        }
        is JSONArray -> (0 until value.length()).joinToString(",", "[", "]") { canonical(value.get(it)) }
        is String -> JSONObject.quote(value)
        is Number -> BigDecimal(value.toString()).stripTrailingZeros().toPlainString()
        is Boolean -> value.toString()
        else -> throw IllegalArgumentException("Invalid action JSON")
    }
}

class CounterLedgerException(val code: String) : IllegalArgumentException(code)

/** Establishing baselines and authorizing corrections are transaction duties.
 * This pure equation never infers an increment from a tracking snapshot. */
object CounterLedger {
    fun project(baseline: JSONObject, events: JSONArray): JSONObject {
        fun fail(code: String): Nothing = throw CounterLedgerException(code)
        val counts = baseline.optJSONObject("counts") ?: fail("INVALID_BASELINE")
        val evidence = baseline.optJSONArray("evidenceIds") ?: fail("INVALID_BASELINE")
        val plans = ActionJson.integer(counts.opt("planViewCount")) ?: fail("INVALID_BASELINE")
        val postpones = ActionJson.integer(counts.opt("dailyPostponeCount")) ?: fail("INVALID_BASELINE")
        if (ActionJson.integer(baseline.opt("schemaVersion")) != 1L
            || !ActionJson.identity(baseline.opt("baselineId")) || !ActionJson.identity(baseline.opt("accountId"))
            || !ActionJson.day(baseline.opt("day")) || plans < 0 || postpones < 0) fail("INVALID_BASELINE")
        val evidenceIds = mutableSetOf<String>()
        for (index in 0 until evidence.length()) {
            val id = evidence.opt(index)
            if (!ActionJson.identity(id) || !evidenceIds.add(id as String)) fail("INVALID_BASELINE")
        }
        val totals = mutableMapOf("planViewCount" to BigInteger.valueOf(plans), "dailyPostponeCount" to BigInteger.valueOf(postpones))
        val seen = mutableMapOf<String, String>()
        for (index in 0 until events.length()) {
            val event = events.optJSONObject(index) ?: fail("INVALID_DELTA")
            val actor = event.opt("actorId")
            val zone = event.opt("timeZone")
            val delta = ActionJson.integer(event.opt("delta")) ?: fail("INVALID_DELTA")
            if (ActionJson.integer(event.opt("schemaVersion")) != 1L
                || !ActionJson.identity(event.opt("actionId")) || !ActionJson.identity(event.opt("accountId"))
                || actor !is String || actor.length !in 1..240 || !ActionJson.day(event.opt("day"))
                || zone !is String || !Regex("^[A-Za-z0-9_+./-]{1,128}$").matches(zone)
                || event.opt("counter") !in totals.keys || delta == 0L || !ActionJson.instant(event.opt("capturedAt"))
                || !event.has("businessActionId") || !event.has("correctionOf")
                || (!event.isNull("businessActionId") && !ActionJson.identity(event.opt("businessActionId")))
                || (if (event.isNull("correctionOf")) delta != 1L else !ActionJson.identity(event.opt("correctionOf")))) fail("INVALID_DELTA")
            if (event.getString("accountId") != baseline.getString("accountId")) fail("SCOPE_MISMATCH")
            val actionId = event.getString("actionId")
            if (actionId in evidenceIds) fail("IDENTITY_MISMATCH")
            val fingerprint = ActionJson.canonical(event)
            val previous = seen[actionId]
            if (previous != null && previous != fingerprint) fail("IDENTITY_MISMATCH")
            if (previous != null) continue
            seen[actionId] = fingerprint
            if (event.getString("day") != baseline.getString("day")) continue
            val counter = event.getString("counter")
            totals[counter] = totals.getValue(counter) + BigInteger.valueOf(delta)
        }
        val projection = JSONObject(counts.toString())
        for ((counter, total) in totals) {
            if (total < BigInteger.ZERO || total > BigInteger.valueOf(ActionJson.MAX_SAFE_INTEGER)) fail("RANGE")
            projection.put(counter, total.toLong())
        }
        return projection
    }
}

File: macos-native/GoalflowMac/Domain/CounterLedger.swift
import Foundation
import CoreFoundation

/// Shared JSON validation for newly captured causal actions. Legacy transport
/// payloads and their original timestamp precision are not rewritten here.
enum ActionJSON {
    static let maxSafeInteger = 9_007_199_254_740_991
    static func identity(_ value: Any?) -> Bool {
        guard let value = value as? String else { return false }
        return value.range(of: "^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$", options: .regularExpression) != nil
    }
    static func integer(_ value: Any?) -> Int? {
        guard let value = value as? NSNumber, CFGetTypeID(value) != CFBooleanGetTypeID(),
              value.doubleValue.isFinite, abs(value.doubleValue) <= Double(maxSafeInteger),
              value.doubleValue.rounded(.towardZero) == value.doubleValue else { return nil }
        return value.intValue
    }
    static func instant(_ value: Any?) -> Bool {
        guard let value = value as? String,
              value.range(of: "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{3}Z$", options: .regularExpression) != nil else { return false }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = formatter.date(from: value) else { return false }
        return formatter.string(from: date) == value
    }
    static func day(_ value: Any?) -> Bool {
        guard let value = value as? String,
              value.range(of: "^\\d{4}-\\d{2}-\\d{2}$", options: .regularExpression) != nil else { return false }
        return instant(value + "T00:00:00.000Z")
    }
}

struct CounterLedgerError: Error {
    let code: String
}

/// The transaction establishes the baseline and authorizes correction evidence.
/// This function only projects distinct deltas; snapshots cannot mint events.
enum CounterLedger {
    static func project(baseline: [String: Any], events: [Any]) throws -> [String: Any] {
        func failure(_ code: String) -> CounterLedgerError { CounterLedgerError(code: code) }
        guard ActionJSON.integer(baseline["schemaVersion"]) == 1,
              ActionJSON.identity(baseline["baselineId"]), ActionJSON.identity(baseline["accountId"]),
              ActionJSON.day(baseline["day"]), let counts = baseline["counts"] as? [String: Any],
              let plans = ActionJSON.integer(counts["planViewCount"]), plans >= 0,
              let postpones = ActionJSON.integer(counts["dailyPostponeCount"]), postpones >= 0,
              let evidence = baseline["evidenceIds"] as? [String],
              evidence.allSatisfy({ ActionJSON.identity($0) }), Set(evidence).count == evidence.count else { throw failure("INVALID_BASELINE") }
        var totals = ["planViewCount": Decimal(plans), "dailyPostponeCount": Decimal(postpones)]
        var seen: [String: String] = [:]
        for raw in events {
            guard let event = raw as? [String: Any], ActionJSON.integer(event["schemaVersion"]) == 1,
                  ActionJSON.identity(event["actionId"]), ActionJSON.identity(event["accountId"]),
                  let actor = event["actorId"] as? String, (1...240).contains(actor.utf16.count),
                  ActionJSON.day(event["day"]), let zone = event["timeZone"] as? String,
                  zone.range(of: "^[A-Za-z0-9_+./-]{1,128}$", options: .regularExpression) != nil,
                  let counter = event["counter"] as? String, totals[counter] != nil,
                  let delta = ActionJSON.integer(event["delta"]), delta != 0,
                  ActionJSON.instant(event["capturedAt"]),
                  event["businessActionId"] is NSNull || ActionJSON.identity(event["businessActionId"]),
                  (event["correctionOf"] is NSNull ? delta == 1 : ActionJSON.identity(event["correctionOf"])) else { throw failure("INVALID_DELTA") }
            guard event["accountId"] as? String == baseline["accountId"] as? String else { throw failure("SCOPE_MISMATCH") }
            let actionID = event["actionId"] as! String
            if evidence.contains(actionID) { throw failure("IDENTITY_MISMATCH") }
            let fingerprint = stableJson(event)
            if let previous = seen[actionID] {
                guard previous == fingerprint else { throw failure("IDENTITY_MISMATCH") }
                continue
            }
            seen[actionID] = fingerprint
            if event["day"] as? String != baseline["day"] as? String { continue }
            totals[counter]! += Decimal(delta)
        }
        var projection = counts
        for (counter, total) in totals {
            guard !total.isNaN, total >= 0, total <= Decimal(ActionJSON.maxSafeInteger) else { throw failure("RANGE") }
            projection[counter] = NSDecimalNumber(decimal: total).intValue
        }
        return projection
    }
}
